ASoC: SOF: IPC4 message abstraction and API change

ujfalusi opened the pull request:
Hi,

Create a container struct to encompass the IPC4 message (the header and the optional data part) so that we can simplify the API and make it possible later on to have generic ops among different IPC versions.

Note: I'm not sure about the new msg_bytes/reply_bytes definition for IPC4, I felt that passing 0 as msg_bytes when there is a message (there must be always) but we don't have the optional data, strange, but it can be justified, if it is preferred.

ujfalusi commented on sound/soc/sof/ipc4.c:
```diff
@@ -219,6 +219,15 @@ int sof_ipc4_init_msg_memory(struct snd_sof_dev *sdev)
+	 * The msg_total_size should be passed as msg_bytes for function expecting
+	 * a message size and underlaying code should take this into account.
+	 */
+	sdev->ipc->max_payload_size += sizeof(struct sof_ipc4_msg);
```
I have missed the setup of the reply placeholder, this should be added after line 227:
```c
	ipc4_msg = msg->reply_data;
	ipc4_msg->data_ptr = msg->reply_data + sizeof(struct sof_ipc4_msg);
```
So that the data_ptr points to the memory right after the header.

Or I could have two separate allocations for the message struct then the data section?

RanderWang reviewed (commented):
LGTM

ranj063 commented on sound/soc/sof/ipc4.c:
```diff
@@ -219,6 +219,15 @@ int sof_ipc4_init_msg_memory(struct snd_sof_dev *sdev)
+	 * The msg_total_size should be passed as msg_bytes for function expecting
+	 * a message size and underlaying code should take this into account.
+	 */
+	sdev->ipc->max_payload_size += sizeof(struct sof_ipc4_msg);
```
I dont understand this @ujfalusi . The max payload for me refers to the MAX size of the mailbox. Why would we include the header and extension in it?

ujfalusi replied:
This was exactly my concern.
I guess I can modify things that the msg_bytes is the optional data size which goes via the mailbox, but then the msg_bytes is kind of half backed as we still have the u64 header, which is a message by itself in some cases.

ujfalusi commented:
Changes since v1:
- the size of `struct sof_ipc4_msg` is not counted towards the msg/reply_bytes
- Remove the header and extension members from `struct snd_sof_ipc_msg`
- Use the new ipc4 encapsulation for message reception

ujfalusi commented:
Changes since v2:
- Remove the header parameter from `sof_ipc_tx_message()` and derivatives
- Remove the header and extension members from `struct snd_sof_ipc_msg`

ranj063 commented on sound/soc/sof/ipc4.c:
```diff
@@ -97,6 +97,7 @@ EXPORT_SYMBOL(sof_ipc4_process_reply);
 static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
 			     void *reply_data)
 {
+	struct sof_ipc4_msg *ipc4_msg = msg->reply_data;
```
@ujfalusi I'm not able to appreciate why we need to do this. Maybe I'm missing the point. There's struct sof_ipc_msg that we use to pass the IPC message to be sent. What is the purpose of the sof_ipc4_msg? we dont have a sof_ipc3_msg similar to sof_ipc4_msg isnt it? So why is ipc4 special?

ujfalusi replied:
@ranj063, all ipc3 messages start with `struct sof_ipc_cmd_hdr` followed by the payload, if any. When we receive the reply or notification then the header is behind of the msg->reply_data.
In IPC3 we pass the _whole_ message, including the header via mailbox.

IPC4 on the other hand is a 'split' protocol in this sense. The header is not passed via the mailbox, it is written to registers and the mailbox is only used in case there is additional data for the message.

We could do an asymmetric handling of IPC4, but I don't think it helps the reader by using the `struct sof_ipc4_msg` for TX and spreading out the reply, notification into `struct sof_ipc_msg`

Users of the IPC interface should not be using the `struct sof_ipc_msg` as it is internal to the IPC 'subsystem' of SOF, we should have a simple interface of "send this message and give me the reply here" and on the notification side it should be that, "here, we have received this message".
Ideally only the low level and high level should know how to deal with the pointer, in the middle layer we just pass things.
I consider the `ipc4_tx_wait_done()` here a low(ish) level functionality as it does error checking, but it should be possible to abstract it away to only wait w/o knowing anything about the content.

ujfalusi replied:
Oh, I see now.
The RX for notification is handled completely differently than the reply in IPC3. The reply is copied to reply_data while the notification is not, the IPC3 RX function itself reads the message from the mailbox (first the header only, then the whole message when it knows the size).

Hrm, It might be better to rename the `reply_data` and `reply_size` to `rx_data` and `rx_size` and use it for both reply and notifications/events.
We are serializing things, so while a reply is handled it is guarantied that a notification will not change the `rx_data`.

That would make it a bit more well defined and we would save on `kmalloc()` for notifications as a bonus?

ranj063 commented on sound/soc/sof/sof-priv.h:
```diff
@@ -369,8 +369,6 @@ struct snd_sof_mailbox {
 /* IPC message descriptor for host <-> DSP IO */
 struct snd_sof_ipc_msg {
 	/* message data */
-	u32 header;
```
typo in commit underlying

ranj063 reviewed (commented):
@ujfalusi should these patches be fixups? the ipc4 stuff is not upstream yet. Maybe thats why I am not able to appreciate why we're doing this

ujfalusi commented:
Changes since v3:
- Move the `struct sof_ipc4_msg` definition to global header (`include/sound/sof/ipc4/header.h`) so it can be used outside of the core (not via sof-priv.h)

ujfalusi commented:
Changes since v4:
- Fixed the tx error handling to print the tx message's primary and extension.

ranj063 commented on sound/soc/sof/ipc4.c:
```diff
@@ -97,6 +97,7 @@ EXPORT_SYMBOL(sof_ipc4_process_reply);
 static int ipc4_tx_wait_done(struct snd_sof_ipc *ipc, struct snd_sof_ipc_msg *msg,
 			     void *reply_data)
 {
+	struct sof_ipc4_msg *ipc4_msg = msg->msg_data;
```
typo in commit message "ip4" and this should be a fixup no?

ujfalusi replied:
Hm, I think it can not be a fixup as there might have been too many changes in ipc4 since it's introduction.

ranj063 commented on sound/soc/sof/debug.c:
```diff
@@ -236,7 +236,7 @@ static int memory_info_update(struct snd_sof_dev *sdev, char *buf, size_t buff_s
 	}
 
-	ret = sof_ipc_tx_message(sdev->ipc, msg.cmd, &msg, msg.size, reply, SOF_IPC_MSG_MAX_SIZE);
+	ret = sof_ipc_tx_message(sdev->ipc, &msg, msg.size, reply, SOF_IPC_MSG_MAX_SIZE);
```
typo in commit message "dropped from as"

ujfalusi commented:
Changes since v5:
- Fixed the typos in commit messages pointed out by @ranj063

bardliao reviewed (dismissed):
LGTM

RanderWang commented on sound/soc/sof/ipc4.c:
```diff
@@ -183,18 +185,19 @@ int sof_ipc4_tx_message(struct snd_sof_ipc *ipc, u32 header, u32 extension,
 		/* check for FW boot completion */
 		if (sdev->fw_state == SOF_FW_BOOT_IN_PROGRESS) {
-			err = sof_ops(sdev)->fw_ready(sdev, msg);
+			err = sof_ops(sdev)->fw_ready(sdev, ipc4_msg->primary);
```
Can you help to add debug log for default case to notify unsupported reply ? thanks!

ujfalusi replied:
Right, should I add a new patch on top or a new PR? We can only print the primary and extension for now, but we should have similar logging as with IPC3.

ujfalusi replied:
@RanderWang, something like this?
https://github.com/ujfalusi/sof-linux/commit/41ebbc9c34242cb19432765be137f564038671c0

RanderWang replied:
sure

RanderWang replied:
you can create a new PR if you like

RanderWang reviewed (approved):
LGTM
